lib/elixir/lib/calendar/iso.ex

```diff
@@ -1455,6 +1455,134 @@ defmodule Calendar.ISO do
     {days, {@parts_per_day - 1, @parts_per_day}}
   end
 
+  @doc """
+  Shifts date by Calendar.Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_date(2016, 1, 3, month: 2)
+      {2016, 3, 3}
+      iex> Calendar.ISO.shift_date(2016, 2, 29, month: 1)
+      {2016, 3, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, month: 1)
+      {2016, 2, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, year: 4, day: 1)
+      {2020, 2, 1}
+  """
+  @spec shift_date(year(), month(), day(), [Calendar.Duration.duration_unit()]) ::
+          {year, month, day}
+  @impl true
+  def shift_date(year, month, day, duration_units) do
+    {year, month, day, _, _, _, _} =
+      shift_naive_datetime(year, month, day, 0, 0, 0, {0, 0}, duration_units)
+
+    {year, month, day}
+  end
+
+  @doc """
+  Shifts naive datetime by Calendar.Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, hour: 1)
+      {2016, 1, 3, 1, 0, 0, {0, 0}}
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, hour: 30)
+      {2016, 1, 4, 6, 0, 0, {0, 0}}
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, microsecond: 100)
+      {2016, 1, 3, 0, 0, 0, {100, 6}}
+  """
+  @spec shift_naive_datetime(year, month, day, hour, minute, second, microsecond, [
+          Calendar.Duration.duration_unit()
+        ]) :: {year, month, day, hour, minute, second, microsecond}
+  @impl true
+  def shift_naive_datetime(year, month, day, hour, minute, second, microsecond, duration_units) do
+    Enum.reduce(duration_units, {year, month, day, hour, minute, second, microsecond}, fn
+      {_opt, 0}, naive_datetime -> naive_datetime
+      {:year, value}, naive_datetime -> shift_years(naive_datetime, value)
+      {:month, value}, naive_datetime -> shift_months(naive_datetime, value)
+      {:week, value}, naive_datetime -> shift_weeks(naive_datetime, value)
+      {time_unit, value}, naive_datetime -> shift_numerical(naive_datetime, value, time_unit)
+    end)
+  end
+
+  def shift_numerical(naive_datetime, value, :day) do
```

Comment:
These should be private :)

Reply:
I was wondering if we could reuse these in `add/3` but that's probably nothing worth considering to consolidate at this point

Reply:
It might actually, so that's a good idea to hold in the back of your head. Especially if you do the refactoring I suggest, we will only have two shift functions: `shift_time_unit` and `shift_month`, and the `shift_time_unit` we can reuse!